shell/platform/android/io/flutter/embedding/engine/FlutterEngineGroup.java

```diff
@@ -117,6 +117,28 @@ public FlutterEngine createAndRunEngine(
       @NonNull Context context,
       @Nullable DartEntrypoint dartEntrypoint,
       @Nullable String initialRoute) {
+    return createAndRunEngine(context, dartEntrypoint, initialRoute, null);
+  }
+
+  /**
+   * Creates a {@link io.flutter.embedding.engine.FlutterEngine} in this group and run its {@link
+   * io.flutter.embedding.engine.dart.DartExecutor} with the specified {@link DartEntrypoint}, the
+   * specified {@code initialRoute} and the {@code dartEntrypointArgs}.
+   *
+   * <p>If no prior {@link io.flutter.embedding.engine.FlutterEngine} were created in this group,
+   * the initialization cost will be slightly higher than subsequent engines. The very first {@link
+   * io.flutter.embedding.engine.FlutterEngine} created per program, regardless of
+   * FlutterEngineGroup, also incurs the Dart VM creation time.
+   *
+   * <p>Subsequent engine creations will share resources with existing engines. However, if all
+   * existing engines were {@link io.flutter.embedding.engine.FlutterEngine#destroy()}ed, the next
+   * engine created will recreate its dependencies.
+   */
+  public FlutterEngine createAndRunEngine(
```

Comment:
3 arguments is kind of my limit, should we use a builder here instead?

```java
FlutterEngine engine = 
  engineGroup.builder()
    .setContext(context)
    .setDartEntryPoint(entryPoint)
    .setInitialRoute(initialRoute)
    .setDartEntrypointArgs(args)
    .build();
```

Reply:
Builder mode is generally used to replace the constructor when there are too many parameters. Maybe we can refactor the API to put all the parameters into `Options` like `BinaryMessenger.makeBackgroundTaskQueue`.

https://github.com/flutter/engine/blob/5fcc59bbc290b9dd8625f2f9a3d5ac7f3f52da85/shell/platform/android/io/flutter/plugin/common/BinaryMessenger.java#L78

``` java
FlutterEngineGroup.Options options = new FlutterEngineGroup.Options()
                                      .setDartEntryPoint(entryPoint)
                                      .setInitialRoute(initialRoute)
                                      .setDartEntrypointArgs(args);

FlutterEngine engine = engineGroup.createAndRunEngine(context, options);
```

If this makes sense to you, on the iOS side, should we do the same thing? It might look like this
``` objc
FlutterEngineGroupOptions* options = [FlutterEngineGroupOptions new];
options.entrypoint = entrypoint;
options.libraryURI = libraryURI;
options.initialRoute = initialRoute;
options.entrypointArgs = entrypointArgs;

FlutterEngine* engine = [engineGroup makeEngineWithOptions:options];

```

Reply:
(I have implemented the logic , here is the code for Android : https://github.com/flutter/engine/pull/29096/commits/ca0ee636537b51485587dc711871c25ef28ac22a , here is the code for iOS : https://github.com/flutter/engine/pull/29096/commits/8d8366997b451a50fa02c6338f55228c81979c29 )

Reply:
I guess the PR https://github.com/flutter/engine/pull/29181 can also benefit from this change.
before:
``` objc
// FlutterViewController

- (instancetype)initWithEngineGroup:(FlutterEngineGroup*)engineGroup
                            nibName:(nullable NSString*)nibName
                             bundle:(nullable NSBundle*)nibBundle
                         entrypoint:(nullable NSString*)entrypoint
                         libraryURI:(nullable NSString*)libraryURI
                        initalRoute:(nullable NSString*)initalRoute NS_DESIGNATED_INITIALIZER;
```

after:
``` objc
// FlutterViewController

- (instancetype)initWithEngineGroup:(FlutterEngineGroup*)engineGroup
                            nibName:(nullable NSString*)nibName
                             bundle:(nullable NSBundle*)nibBundle
                            options:(nullable FlutterEngineGroupOptions*)options NS_DESIGNATED_INITIALIZER;
```

Reply:
Done